src/Storages/StorageAggregatingMemory.cpp

```diff
@@ -0,0 +1,484 @@
+        for (size_t i = 0; i < data.params.keys_size; ++i)
+        {
+            if (data.header.getByPosition(i).name == ident->name())
+            {
+                key_position = i;
+                break;
+            }
+        }
+
+        /// This identifier is not a key.
+        if (key_position == -1)
+            return;
+
+        /// function->name == "equals"
+        if (const auto * literal = value->as<ASTLiteral>())
+        {
+            auto column_type = data.header.getByPosition(key_position).type;
+
+            auto converted_field = convertFieldToType(literal->value, *column_type);
+            if (!converted_field.isNull() && data.raw_key_columns[key_position]->empty())
+            {
+                data.raw_key_columns[key_position]->insert(converted_field);
+            }
+        }
+    }
+};
+
+using AggregationKeyVisitor = AggregationKeyMatcher::Visitor;
+
+std::optional<AggregationKeyVisitor::Data> getFilterKeys(const Aggregator::Params & params, const SelectQueryInfo & query_info)
+{
+    if (params.keys.empty())
+        return {};
+
+    const auto & select = query_info.query->as<ASTSelectQuery &>();
+    ASTPtr where = select.where();
+    if (!where)
+        return {};
+
+    AggregationKeyVisitor::Data data(params);
+    AggregationKeyVisitor(data).visit(where);
+
+    if (data.failed || !data.hasAllKeys())
+        return {};
+
+    return data;
+}
+
+void executeExpression(Pipe & pipe, const ActionsDAGPtr & expression)
+{
+    if (!expression)
+    {
+        return;
+    }
+
+    auto expression_actions = std::make_shared<ExpressionActions>(expression);
+
+    pipe.addSimpleTransform([expression_actions](const Block & header)
+    {
+        return std::make_shared<ExpressionTransform>(header, expression_actions);
+    });
+}
+
+/// AggregatingOutputStream is used to feed data into Aggregator.
+class AggregatingOutputStream : public IBlockOutputStream
+{
+public:
+    AggregatingOutputStream(StorageAggregatingMemory & storage_, const StorageMetadataPtr & metadata_snapshot_, ContextPtr context_)
+        : storage(storage_),
+          metadata_snapshot(metadata_snapshot_),
+          context(context_),
+          variants(*(storage.many_data)->variants[0]),
+          key_columns(storage.aggregator_transform->params.keys_size),
+          aggregate_columns(storage.aggregator_transform->params.aggregates_size)
+    {
+        expression_actions = std::make_shared<ExpressionActions>(storage.analysis_result.before_aggregation);
+    }
+
+    // OutputStream structure is same as source (before aggregation).
+    Block getHeader() const override { return storage.src_block_header; }
+
+    void write(const Block & block) override
+    {
+        storage.src_metadata_snapshot->check(block, true);
+
+        Block block_for_aggregation(block);
+        expression_actions->execute(block_for_aggregation);
+
+        bool no_more_keys = false;
+        storage.aggregator_transform->aggregator.executeOnBlock(
```

Comment:
Looks like race between may writes (and between read and write)

Reply:
Code has been changed, but I think this issue still exists. Do you think using `shared_mutex` for aggregator updates and reads will be fine for the fix?